internal/dag/dag.go

```diff
@@ -410,7 +410,7 @@ func (d *DAG) processIngressRoute(ir *ingressroutev1.IngressRoute, prefixMatch s
 			for _, s := range route.Services {
 				m := meta{name: s.Name, namespace: ir.Namespace}
 				if svc := service(m, intstr.FromInt(s.Port)); svc != nil {
-					r.addService(svc)
+					r.addService(svc, s.HealthCheck, s.Strategy)
```

Comment:
I think this is not right, but we'll live with it for now.

Fundamentally two services with different ports are different, from the dag point of view. Therefore two services with the same port but different healthcheck or lb strategies are different. 

Either

a. we place those differences on the route object, which matches where they came from in the ingressroute document
b. we make healthcheck params and lb strategy part of the key for the _services map--so extend portmeta to include more parameters.

This is fine to commit for beta.1, please raise an issue to address this for 0.6 final.

Thanks

Reply:
If we put this on the Route Object, how would you move that into the CDS update time?

Reply:
Also confused about your comment regarding different services, this PR maintains different healthchecks or lb strat per service.

Reply:
> Also confused about your comment regarding different services, this PR maintains different healthchecks or lb strat per service.

Sorry, I probably wasn't very clear.

At the moment the _services (terrible name, sorry) map is keyed by a tuple of namespace/name/port; so the same k8s service will result in two different service objects in the dag if a route points to two different ports on the same service.  https://github.com/heptio/contour/blob/master/internal/contour/cluster_test.go#L220

Arguably lbstrat and healthcheck params are properties of the service namespace/name/port tuple. I think I can argue that for the latter, why would the a service have two different ways of checking its health depending on which routes point to it. I could probably make the same argument for lbstrat. If that argument holds then those properties should be defined on the k8s service object via annotations, and incorporated into the dag.Service object -- which means they are _not_ part of the _service's map key.

However, because we currently define lbstrat and hc on the ingressroute object, it is possible to have more than one way to point to a single k8s service with different values for those parameters. So, what we should probably do in that case is make lbstrat and healthcheck params properties of the route so when the cds visitor is building the cds table it knows to emit distinct cluster values for each cluster depending on the path it took to reach that vertex.

However, this argument is problematic because if a route has multiple service children, does that mean we need to store an array of healthcheck and lbstrat configurations on the Route object? The ingressroute document lets us define a different hc and lbstrat per service so we need to store and map an array of hc/lbstrat values in the route object to the route objects children. 

This is the same situation I found myself in when doing RDS, a ingressroute Route can have multiple service name / port pairs. But the child of a dag.Route was a k8s Service object, which can have multiple ports defined. Said another way, the ingressroute Route servicename field gave us the k8s service object to wrap in a dag.Service and insert into the graph, but it didn't contain enough information for the routeVisitor to recover the port details of the child dag.Service. The first solution I tried was to create a map of ports to children in the route object, but that was brittle. The better solution was to create a distinct dag.Service for a service name and port combination. This also matched the envoy Cluster model.

In summary, what is here will work for beta.1, but we need to revisit it before 0.6 final because I think that these parameters belong on the service document, which is problematic for the discoverers.